### PR TITLE
Automated cherry pick of #125935: Terminate restartable init containers ignoring not-started containers

### PR DESCRIPTION
**Pull request by gjkim42**
Cherry pick of #125935 on release-1.29.

#125935: Terminate restartable init containers ignoring not-started containers

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
Fixed a bug that init containers with `Always` restartPolicy may not terminate gracefully if the pod hasn't initialized yet.
```

### CODE REVIEW

**Comment by gjkim42** [comment]
/cc @SergeyKanzhelev 
/assign @yujuhong 

This backports #125935 to release-1.29.

**Comment by saschagrunert** [comment]
Waiting for approval on https://github.com/kubernetes/kubernetes/pull/126332 and https://github.com/kubernetes/kubernetes/pull/126331

**Comment by SergeyKanzhelev** [comment]
> Waiting for approval on #126332 and #126331

I pinged @yujuhong, I think one of PRs was just lost among notifications

**Comment by gjkim42** [comment]
/assign @saschagrunert 
#126331 is approved by node approval now. Could you PTAL?

**Review by saschagrunert — APPROVED** [review]

**Pull request merged**
